fix(compiler-cli): ensure FS case-sensitivity is correct

petebacondarwin opened the pull request:
The `NodeJSFileSystem.isCaseSensitive()` method was computing
completely the opposite value for its result.

Also the `NgtscCompilerHost.getCanonicalFileName()` method was
not actually calling `useCaseSensitiveFileNames()` so it was always
assuming that the filenames were case sensitive and not actually
canonicalizing them. Instead just returning the original.

Once these two bugs were fixed, a number of issues with the rest of the
handling of case-insensitive files in compiler-cli and tests were uncovered.

This commit also fixes these issues. Here are some further notes:

The TS compiler canonicalizes filenames internally for lookups etc. So
if we are also doing file lookups we need to use the same canonicalized
forms. This is achieved by injecting the `ts.CompilerHost` where needed
and calling `getCanonicalFileName()`. For example, when computing the
`rootDirs` in the `LogicalFileSystem`.

Some tests were checking that particular source files were being exposed
by comparing the `ts.SourceFile.fileName` property with some non-canonical
filename. In some cases this would fail. Instead these comparisons are
achieved by directly comparing the SourceFile objects, if known, or getting
the SourceFile object from a non-canonical filename from the program, by
calling `program.getSourceFile(nonCanonicalFileName)`.

The `MockFileSystemWindows` class was not considering lower case drive names
when computing if a path is rooted (i.e. absolute).

The `NodeJSFileSystem.isCaseSensitive()` method was relying upon its `exists()`
method. But this itself relies upon the `isCaseSensitive()` to determine
if a file exists. So there was an infinite recursion. This is solved by
actually using the underlying real `fs.existsSync` to compute whether the
file system is case sensitive.

---

This was tested to work on Windows via #36873. See https://app.circleci.com/pipelines/github/angular/angular/14310/workflows/b347d457-f0ce-4d54-afbe-413a1c481003

AndrewKushnir reviewed (approved):
LGTM 👍

(CI errors look legit, so I've added the "cleanup" label)

petebacondarwin commented:
Indeed this fix has revealed a bunch of issues with our Windows support!!

AndrewKushnir commented:
@petebacondarwin just to be safe (and make sure everything works as expected on Windows) it might make sense to run `test_win` and `test_ivy_aot_win` CircleCI jobs. You can enable them in a separate commit in this PR (and revert it later) or create a separate PR (this is what Greg usually does for Bazel-relaated changes), and enable the mentioned CircleCI jobs [for example as shown here](https://github.com/angular/angular/pull/36088/files?file-filters%5B%5D=.yml#diff-1d37e48f9ceff6d8030570cd36286a61L876).

petebacondarwin commented:
Yes definitely

AndrewKushnir commented on packages/compiler-cli/src/ngtsc/file_system/src/logical.ts:
```diff
@@ -76,11 +79,13 @@ export class LogicalFileSystem {
       for (const rootDir of this.rootDirs) {
-        if (physicalFile.startsWith(rootDir)) {
-          logicalFile = this.createLogicalProjectPath(physicalFile, rootDir);
+        if (!relative(rootDir, canonicalFilePath).startsWith('..')) {
```
It looks like there is a similar check in this file that uses `startsWith('../')` condition. I'd propose extracting it into a separate utility function, which would also serve as extra documentation.

petebacondarwin replied:
Actually the two uses of `startsWith('..')` cannot easily be combined into a single function. I will change this usage to call a function.

AndrewKushnir commented on packages/compiler-cli/src/ngtsc/file_system/src/node_js_file_system.ts:
```diff
@@ -68,7 +68,7 @@ export class NodeJSFileSystem implements FileSystem {
   isCaseSensitive(): boolean {
     if (this._caseSensitive === undefined) {
-      this._caseSensitive = this.exists(togglePathCase(__filename));
+      this._caseSensitive = !fs.existsSync(togglePathCase(__filename));
```
There is a note on this change in PR description, but it might make sense to put it here as well (as an inline comment), since using `this.exists` seems like a natural thing to do, but switching it back will result in errors.

AndrewKushnir commented on packages/compiler-cli/src/ngtsc/file_system/testing/src/mock_file_system_windows.ts:
```diff
@@ -34,14 +34,14 @@ export class MockFileSystemWindows extends MockFileSystem {
 
   isRooted(path: string): boolean {
-    return /^([A-Z]:)?([\\\/]|$)/i.test(path);
+    return /^([A-Za-z]:)?([\\\/]|$)/i.test(path);
```
Just curious if this change is needed, since the RegExp has the `i` flag?

petebacondarwin replied:
Doh! Of course.

AndrewKushnir reviewed (approved):
LGTM, just left a couple nits.

IgorMinar commented:
@petebacondarwin can you please rebase this? thanks!

petebacondarwin committed:
fix(compiler-cli): fix case-sensitivity issues in NgtscCompilerHost

The `getCanonicalFileName()` method was not actually
calling the  `useCaseSensitiveFileNames()` method. So
it always returned a case-sensitive canonical filename.

petebacondarwin committed:
fix(compiler-cli): ensure LogicalFileSystem handles case-sensitivity

The `LogicalFileSystem` was not taking into account the
case-sensitivity of the file-system when caching logical
file paths.

petebacondarwin committed:
fix(compiler-cli): ensure `getRootDirs()` handles case-sensitivity

Previously the `getRootDirs()` function was not converting
the root directory paths to their canonical form, which can
cause problems on case-insensitive file-systems.

petebacondarwin committed:
fix(compiler-cli): `isCaseSensitive()` returns correct value

Previously this method was returning the exact opposite value
than the correct one.
Also, calling `this.exists()` causes an infinite recursions,
so the actual file-system `fs.existsSync()` method is used
to ascertain the case-sensitivity of the file-system.

petebacondarwin committed:
fix(compiler-cli): ensure `MockFileSystem` handles case-sensitivity

Previously this class used the file passed in directly to look up files in the
in-memory mock file-system. But this doesn't match the behaviour of
case-insensitive file-systems. Now the look up is done on the canonical
file paths.

petebacondarwin committed:
fix(compiler-cli): normalize mock Windows file paths correctly

Since the `MockFileSystemWindows` is case-insensitive, any
drive path that must be added to a normalized path should be lower
case to make the path canonical.

petebacondarwin committed:
fix(compiler-cli): use CompilerHost to ensure canonical file paths

The type checking infrastrure uses file-paths that may come from the
TS compiler. Such paths will have been canonicalized, and so the type
checking classes must also canonicalize paths when matching.

petebacondarwin committed:
test(compiler-cli): ensure indexer tests are not brittle to case-sensitivity

These tests were matching file-paths against what is retrieved from the
TS compiler. But the TS compiler paths have been canonicalised, so the
tests were brittle on case-insensitive file-systems.

petebacondarwin committed:
test(compiler-cli): ensure partial-evaluator tests are not brittle to case-sensitivity

These tests were matching file-paths against what is retrieved from the
TS compiler. But the TS compiler paths have been canonicalised, so the
tests were brittle on case-insensitive file-systems.

petebacondarwin committed:
test(compiler-cli): ensure reflection tests are not brittle to case-sensitivity

These tests were matching file-paths against what is retrieved from the
TS compiler. But the TS compiler paths have been canonicalised, so the
tests were brittle on case-insensitive file-systems.

petebacondarwin commented:
Rebased...

alxhub commented:
@petebacondarwin this didn't merge cleanly to patch :( please open a separate PR if you feel like it should be merged there.
